### ISSUE TITLE
Remove setOnPictureTakenHandler and instead use takePicture onSuccess

### ISSUE THREAD

**Issue by westonganger** [cordova-plugin-camera-preview/cordova-plugin-camera-preview#217]
I am going to be changing the API to remove setOnPictureTakenHandler and instead use the onSuccess function of takePicture. This is a more common API style that feels more natural.

If anyone has any concerns with this, let me know why

**Comment by mangel1196** [comment]
Hi.

Just let us know when the change is over

Thanks

**Comment by westonganger** [comment]
This has now been changed. For the new instructions see readme for `takePicture`.

**Comment by 4UForever** [comment]
Hi.
I'm implement follow this doc by ionic framework 2
[https://ionicframework.com/docs/v2/native/camerapreview/](url)
But now the plugin doesn't work.
When I use new takePicture method

> CameraPreview.takePicture(function(base64PictureData){
>      console.log(base64PictureData);
> });

The console has show error like this.

> Typescript Error
Argument of type '(base64PictureData: any) => void' is not assignable to parameter of type 'CameraPreviewSize'. Property 'maxWidth' is missing in type '(base64PictureData: any) => void'.

I think it because the doc doesn't update and also node_modules package doesn't update relate to this package url
[https://www.npmjs.com/package/cordova-camera-preview](url)

I found the file in **/node_modules/ionic-native/dist/es5/plugins/camera-preview.js** and **camera-preview.d.ts** still have method "setOnPictureTakenHandler" and method "takePicture" still have only one parameter (size?: CameraPreviewSize).

So, Could you please update node_modules package will be thankful.
Thanks.

**Comment by westonganger** [comment]
Read the readme for master and use the current install instructions. We are drastically improving this so there wont be a new NPM release until its stable.

**Comment by jlongman** [comment]
So I'm testing this, I'm at 946863c2 - which should be the `CFAutoRelease` commit and includes the `takePicture` change, and I note that if I tap on the preview 3 times I will get 3 successive pictures taken.  There's no debouncing, or event amalgamation.

I don't remember this behaviour before - but maybe because old code (going back months) took smaller pictures by default, maybe it completed the upload before I tapped again and it wouldn't look like a debounce problem.  So, is this something introduced by this change?  Is there anything inherently different about the `setOnTakePictureHandler` vs `takePicture` success that would make this change? 

(Apologies - I'm less familiar with the JavaScript side so this may be a dumb question).

**Comment by jlongman** [comment]
Also, and I don't remember this either but I'm very human, if I do a `location.reload()` or a hot code push, then a picture will be taken.

**Comment by westonganger** [comment]
I am curious why it would take a photo on reload. I think that is likely your own projects code making a takePicture call.

As for the 3 taps == 3 photos, we need to introduce a lock as described in #221

**Comment by jlongman** [comment]
A lock sounds good, I wasn't clear how safe it would be in JavaScript (but that's my weakness in JS showing). 

My thought re:`location.reload()` is that previously `setOnTakePictureHandler` being called a second time would just set the handler again, does calling `takePicture` twice - twice since the `reload()` will re-init the app and set the handler as the `success` function again - not take a picture since it calls `invokeTakePicture` in the iOS code? (and if so, why doesn't it the first time, or does it and I just don't have my logging setup yet?)

**Comment by westonganger** [comment]
Your second paragraph was quite hard to follow. Calling the takePicture twice fast, which should not occur if we have a lock, would overwrite the first takePicture onSuccess callback with the second takePicture onSuccess callback. Basically the camera preview holds a reference to last takePicture callbacks.

**Comment by jlongman** [comment]
To be clear, as I edited above, I use the `tap` on the preview as setup in the `startCamera`, so I never previously called `takePicture`, only `setOnTakePictureHandler`.  Now I call `takePicture` and it takes a picture when I set the handler.  I don't want to take a picture, I only want the handler set.  As I see it this was the distinction between the two calls.  

Now I set my handler and when I `location.reload()` or even just load the app maybe I take a picture.

**Comment by westonganger** [comment]
Couldnt you just do:

```javascript
var onPictureTakenHandler = function(base64ImageData){
  // do work
};

CameraPreview.takePicture(onPictureTakenHandler);
```

**Comment by westonganger** [comment]
Okay so the tap to take a picture might require the setOnPictureTakenHandler method. Maybe its better to have a `setTapPhotoHandler` instead?

**Comment by jlongman** [comment]
(To be pedantic for future readers, `CameraPreview.takePicture` calls `invokeTakePicture` which will take a picture and not just register the handler so the 'tap preview' convenience code as provided in `startCamera` needs different handling)

Will I want/not want to have 'tap' handler and 'takePicture' handler the same or different?

Personally I would revert to the previous API because my use case matches it but I also feel like the handler is an observer to the picture being taken.  The only question is whether the the tapHandler and takePicture handler are kept distinct, I guess.

(The `setFlashMode` and `startCamera` "race condition" I mention in #167 makes me see why you might be doing this - no race between the handler being set and the picture being taken.)

**Comment by westonganger** [comment]
So after digging into this there are problems with the new approach because of the tapPhoto. We will probably need to change the following things:

1. We need to revert to using `setOnPictureTakenHandler` so that the callback is also set for tapPhoto.
2. We need to move the picture options out of `takePicture` and create a new method `setTakePictureOptions` so that the options are also set for tapPhoto

I dont think its the nicest API but the native implementation with threads points to this solution. If anyone has any comments or improvements to this, let me know.
